src/TenantCloud/APIVersioning/APIVersioningServiceProvider.php

```diff
@@ -2,41 +2,60 @@
 
 namespace TenantCloud\APIVersioning;
 
-use Illuminate\Contracts\Foundation\Application;
-use Illuminate\Http\Request;
 use Illuminate\Routing\Contracts\ControllerDispatcher;
 use Illuminate\Routing\Route;
 use Illuminate\Support\ServiceProvider;
-use TenantCloud\APIVersioning\Version\VersionFromHeader;
+use TenantCloud\APIVersioning\Constraint\BuiltInStringConstraintParser;
+use TenantCloud\APIVersioning\Constraint\ConstraintChecker;
+use TenantCloud\APIVersioning\Constraint\SemanticConstraintChecker;
+use TenantCloud\APIVersioning\Constraint\StringConstraintParser;
+use TenantCloud\APIVersioning\Version\RequestHeaderVersionParser;
+use TenantCloud\APIVersioning\Version\RequestVersionParser;
+use TenantCloud\APIVersioning\Version\SemanticVersionParser;
 use TenantCloud\APIVersioning\Version\VersionParser;
 
 class APIVersioningServiceProvider extends ServiceProvider
 {
-	public function boot(): void
-	{
-		$this->publishes([
-			__DIR__ . '/config/api-versioning.php' => config_path('api-versioning.php'),
-		]);
-
-		// Versioned API mixin
-		Route::mixin(new RouteVersionMixin());
-	}
-
 	public function register(): void
 	{
 		parent::register();
 
-		$this->mergeConfigFrom(__DIR__ . '/config/api-versioning.php', 'api-versioning');
-
 		// Bind laravel controller dispatcher for our custom versioned dispatcher.
 		$this->app->bind(
 			ControllerDispatcher::class,
 			VersionControllerDispatcher::class
 		);
 
+		// Bind parser from request header as default.
+		$this->app->singleton(
+			RequestVersionParser::class,
+			RequestHeaderVersionParser::class
+		);
+
+		$this->app->singleton(
+			ConstraintChecker::class,
+			SemanticConstraintChecker::class
+		);
+
 		$this->app->singleton(
 			VersionParser::class,
-			fn (Application $app) => new VersionFromHeader($app->factory(Request::class))
+			SemanticVersionParser::class
+		);
+
+		$this->app->singleton(
+			StringConstraintParser::class,
+			BuiltInStringConstraintParser::class
 		);
 	}
+
+	public function boot(): void
+	{
+		// Versioned API mixin
+		$mixin = new RouteVersionMixin($this->app->make(ConstraintChecker::class));
+
+		Route::macro('versioned', $mixin->versioned());
+		Route::macro('getVersionClassAndMethod', $mixin->getVersionClassAndMethod());
+		Route::macro('isVersionRegister', $mixin->isVersionRegister());
```

Comment:
How is `isVersionRegister` different from `hasRegisteredVersion`? Lexically they have the same meaning

Reply:
isVersionRegister - check is route has at least one matched constraint 
hasRegisteredVersion - will be renamed into hasRegisteredVersions - show if route has any registered constraints